moshi/src/main/java/com/squareup/moshi/JsonValueSource.java

```diff
@@ -0,0 +1,208 @@
+   */
+  private void advanceLimit(long byteCount) throws IOException {
+    while (limit < byteCount) {
+      // If we've finished the JSON object, we're done.
+      if (state == STATE_END_OF_JSON) {
+        return;
+      }
+
+      // If advancing requires more data in the buffer, grow it.
+      if (limit == buffer.size()) {
+        source.require(limit + 1L);
+      }
+
+      // Find the next interesting character for the current state. If the buffer doesn't have one,
+      // then we can read the entire buffer.
+      long index = buffer.indexOfElement(state, limit);
+      if (index == -1L) {
+        limit = buffer.size();
+        continue;
+      }
+
+      byte b = buffer.getByte(index);
+
+      if (state == STATE_JSON) {
+        switch (b) {
+          case '[':
+          case '{':
+            stackSize++;
+            limit = index + 1;
+            break;
+
+          case ']':
+          case '}':
+            stackSize--;
+            if (stackSize == 0) state = STATE_END_OF_JSON;
+            limit = index + 1;
+            break;
+
+          case '\"':
+            state = STATE_DOUBLE_QUOTED;
+            limit = index + 1;
+            break;
+
+          case '\'':
+            state = STATE_SINGLE_QUOTED;
+            limit = index + 1;
+            break;
+
+          case '/':
+            source.require(index + 2);
+            byte b2 = buffer.getByte(index + 1);
+            if (b2 == '/') {
+              state = STATE_END_OF_LINE_COMMENT;
+              limit = index + 2;
+            } else if (b2 == '*') {
+              state = STATE_C_STYLE_COMMENT;
+              limit = index + 2;
+            } else {
+              limit = index + 1;
+            }
+            break;
+
+          case '#':
+            state = STATE_END_OF_LINE_COMMENT;
+            limit = index + 1;
+            break;
+        }
+
+      } else if (state == STATE_SINGLE_QUOTED || state == STATE_DOUBLE_QUOTED) {
+        if (b == '\\') {
+          source.require(index + 2);
+          limit = index + 2;
+        } else {
+          state = (stackSize > 0) ? STATE_JSON : STATE_END_OF_JSON;
+          limit = index + 1;
+        }
+
+      } else if (state == STATE_C_STYLE_COMMENT) {
+        source.require(index + 2);
+        if (buffer.getByte(index + 1) == '/') {
+          limit = index + 2;
+          state = STATE_JSON;
+        } else {
+          limit = index + 1;
+        }
+
+      } else if (state == STATE_END_OF_LINE_COMMENT) {
+        limit = index + 1;
+        state = STATE_JSON;
+
+      } else {
+        throw new AssertionError();
+      }
+    }
+  }
+
+  /**
+   * Discards any remaining JSON data in this source that was left behind after it was closed. It is
+   * an error to call {@link #read} after calling this method.
+   */
+  public void discard() throws IOException {
+    closed = true;
+    while (state != STATE_END_OF_JSON) {
+      advanceLimit(8192);
```

Comment:
Unfortunate that we have this constant in multiple codebases downstream of Okio. I'm not sure how it would even be exposed if we wanted... `Buffer.defaultSegmentSize()`?

Reply:
Yeah, agreed.